src/Storages/MergeTree/MergeTreeIndexMinMax.cpp

```diff
@@ -72,29 +56,53 @@ void MergeTreeIndexGranuleMinMax::deserializeBinary(ReadBuffer & istr)
         const DataTypePtr & type = index_sample_block.getByPosition(i).type;
         auto serialization = type->getDefaultSerialization();
 
-        if (!type->isNullable())
-        {
-            serialization->deserializeBinary(min_val, istr);
-            serialization->deserializeBinary(max_val, istr);
-        }
-        else
+        switch (version)
         {
-            /// NOTE: that this serialization differs from
-            /// IMergeTreeDataPart::MinMaxIndex::load() due to preserve
-            /// backward compatibility.
-            bool is_null;
-            readBinary(is_null, istr);
-            if (!is_null)
-            {
+            case 1:
+                if (!type->isNullable())
+                {
+                    serialization->deserializeBinary(min_val, istr);
+                    serialization->deserializeBinary(max_val, istr);
+                }
+                else
+                {
+                    /// NOTE: that this serialization differs from
+                    /// IMergeTreeDataPart::MinMaxIndex::load() to preserve
+                    /// backward compatibility.
+                    ///
+                    /// But this is deprecated format, so this is OK.
+
+                    bool is_null;
+                    readBinary(is_null, istr);
+                    if (!is_null)
+                    {
+                        serialization->deserializeBinary(min_val, istr);
+                        serialization->deserializeBinary(max_val, istr);
+                    }
+                    else
+                    {
+                        min_val = Null();
```

Comment:
So the difference of the new format is that it can serialize null values for min and max separately? And the old format serialized nulls for both because "one is enough", as the comment said. But now we want more precise nullable support so we want to serialize both.

Reply:
Right.